### PR TITLE
Batch Changes: Replace MultiSelect with wildcard MultiCombobox UI

### PR DESCRIPTION
**Pull request by vovakulikov**
## Background

In this PR https://github.com/sourcegraph/sourcegraph/pull/46362 we added a new wildcard component for MultiCobobox-like UI. MultiSelect (react-select library) provides not so straightforward API and types around this component, it also possible that one of the react-query types slows down TypeScript compiler see related [slack thread ](https://sourcegraph.slack.com/archives/C01C3NCGD40/p1669864732998109?thread_ts=1669859170.510239&cid=C01C3NCGD40)

Also multi select isn't so small library it's 90kb minified, so In this PR we replace this component with our in-house implementation of MultiCombobox UI

Screenshot 2023-01-13 at 22 59 49

## Test plan
- Check Batch changes filters UI 
- Filters should be preserved in URL and also saved in temporary setting store
- Check that with new UI you still can see filters from the old UI version
- Check that list page has right set of entities based on selected filters

## App preview:

- [Web](https://sg-web-vk-replace-multi-select-ui.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.01% (-0.35 kb) | -0.54% (-77.10 kb) 🔽 | -0.66% (-76.75 kb) 🔽 | -0.14% (-1) 🔽 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/vk/replace-multi-select-ui/compare-report.html;tab=live_object) for a full comparison between the commits 1f1ee22 and 1ef9f8b or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by vovakulikov** [comment]
I think we should merge this after the 4.4 release cut just to be sure that we don't break anything about old format filters in URL and temporary settings store.

**Review by courier-new — APPROVED** [review]
Looks good!! I agree about waiting until after 4.4 just in case, but here's my approval. 🙂 I only tested from the Storybook because the client preview app service seems to be down right now, but thank you for your thorough testing.

File: client/web/src/enterprise/batches/list/BatchChangeListFilters.tsx [hunk 1]
```diff
@@ -1,37 +1,85 @@
...
+                            filters.map((filter, index) => (
+                                <MultiComboboxOption key={filter.toString()} value={format(filter)} index={index} />
+                            ))
+                        }
+                    </MultiComboboxList>
+
+                    {suggestions.length === 0 && (
+                        <span className={styles.noFilters}>
+                            {!searchTerm ? (
+                                <>All filters are selected, there are no any other filters</>
```
Comment (courier-new) [inline client/web/src/enterprise/batches/list/BatchChangeListFilters.tsx:75]: Maybe just this is sufficient

```suggestion
                                <>All filters are selected</>
```

File: client/web/src/enterprise/batches/list/BatchChangeListFilters.tsx [hunk 2]
```diff
@@ -1,37 +1,85 @@
...
+/** Returns string with capitalized first letter */
+const format = (filter: BatchChangeState): string => {
+    const str = filter.toString()
 
-export const STATUS_OPTIONS: MultiSelectOption<BatchChangeState>[] = [OPEN_STATUS, DRAFT_STATUS, CLOSED_STATUS]
-// Drafts are a new feature of severside execution that for now should not be shown if
-// execution is not enabled.
-const STATUS_OPTIONS_NO_DRAFTS: MultiSelectOption<BatchChangeState>[] = [OPEN_STATUS, CLOSED_STATUS]
+    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase()
+}
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/web/src/enterprise/batches/list/BatchChangeListFilters.tsx:25]: I slowly forget about all lodash methods, thanks for reminder I will use upperFirst instead

File: client/web/src/enterprise/batches/list/BatchChangeListFilters.tsx [hunk 3]
```diff
@@ -1,37 +1,85 @@
...
+        <Label htmlFor={id} className={classNames(className, styles.root)}>
+            <H4 as={H3} className="mb-0 mr-2">
+                Status
+            </H4>
+
+            <MultiCombobox
+                selectedItems={selectedFilters}
+                getItemName={format}
+                getItemKey={format}
+                onSelectedItemsChange={onFiltersChange}
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply client/web/src/enterprise/batches/list/BatchChangeListFilters.tsx:53]: Aha, agree, but I think this makes sense only in this particular use case, so I will support this in batch change UI (in consumer level). In other cases (for example, in Code Insights add or remove insight UI it makes sense to preserve the value after you pick something from suggestions. 

Maybe I will add these two options in `MultiComboboxInput` UI later (there are other non-trivial problems with bypassing internal reach ui logic around input value prop)

**Comment by vovakulikov** [comment]
@courier-new thanks for your review, all valid and vital comments, I fixed all of them and I think now it's ready to be merged, but let's wait for the 4.4 release cut. 😉

File: client/web/src/enterprise/batches/list/BatchChangeListFilters.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (courier-new) [reply client/web/src/enterprise/batches/list/BatchChangeListFilters.tsx]: That seems reasonable, thanks for adding it for the Batch Changes UI for now. 🙂

**Pull request merged**